### DISCUSSION TITLE
understanding how scene graph works: changing textures

### DISCUSSION THREAD

**Discussion by orija** [google/model-viewer#2367] [Q&A]
hello, I am trying to understand the examples on how to switch textures interactively. It would be wonderful if the documentation was a bit more fleshed out. I am not a beginner, I've used this script for a long time and I love that so many things are possible with it. But it is very hard for someone who only knows lite programming to get into some of these features without some explaination.

What I'd like to do is to have my model change textures when I grab the object. Maybe this is too complicated for a simple answer, but I'm asking you guys anyway. If anyone can help me understand how this can be possible, then thank you.

**Comment by elalish** [comment]
The key is what do you mean by "grab the object"? If you mean interact, you can just use a click event handler to swap variants or textures as shown in the examples: https://modelviewer.dev/examples/scenegraph/

If you mean select a particular object in the scene and then switch the texture just on that object, that's not possible yet, though we are working to flesh out our scene-graph API (part of the sparsity of documentation is that it's still under development).

**Reply by orija** [reply]
thanks for asking. I know what I said is vague. I mean on 'mousedown' to simply swap the texture of the one object in the scene. Sounds very simple. My problem is that I am not sure how to access the material... I see code in the example here https://modelviewer.dev/examples/scenegraph/#swapTextures ...But it applies to a pulldown element. I have a hard time understanding which part of the code I need to switch the texture.
I imagine its something in here:

material.pbrMetallicRoughness[channel].texture.source.setURI(event.target.value);
 
ugh.. this is totally unfamiliar javascript to me.

I have tried so many things. I just want to call a new texture on mousedown, replace it with the original one on mouse up. so imagine, while you rotate the object turns another color, when you let go it goes back to normal.

**Comment by ghost** [comment]
Thanks for the info, yes using the setURI method will change the texture
such as `pbrMetallicRoughness['baseColorTexture']` for the entire model. If
you only wanted to change the texture of one part of a model, for instance,
a wheel on a car model then we don't have that API quite yet but it's being
discussed.

**Reply by dhruvadhia1** [reply]
Actually, one can target specific material through this line by targeting a specific index in an array.

let material = modelViewerTexture.model.materials[0];

**Reply by orija** [reply]
okay so if my texture is called "RedBox-color'.jpg" How would I tell the texture to apply to the model?

let material = modelViewerTexture.model.materials["RedBox-color.jpg"]

sorry for not being better at any of this (-_-)'

**Reply by elalish** [reply]
We really need to make a names-based API to make this easier. For now you have to do it by index. Many models only have a single material, in which case `let material = modelViewerTexture.model.materials[0];` should do it. If that's not the one you want, try `[1]`, `[2]` until you change the one you want. To make this easier, I'd highly recommend finding a program that can export the glTF variants extension so you can pack all the textures in one file and then you can simply select it by name. Make sure to check the console for errors; Chrome dev tools can be very helpful in navigating the object hierarchy you're trying to access.
